students/paul_c/session04/mailroom.py

```diff
@@ -0,0 +1,145 @@
+#!/usr/local/bin/python3
+
+import sys
+from datetime import datetime
+
+donors = {
+    "Anne Ant": [1.00],
+    "Bonnie Bug": [20.00, 10.00],
+    "Chuck Cat": [2.00, 2.00, 2.00],
+    "Donna Dog": [5000.00, 2500.00, 1.50],
+    "Edna Ent": [.30, .50, .10]
+}
+
+email = {
+    "greeting": "\nHello {}\n\n",
+    "body": "We would like to thank you for your generous donation of ${}.\n\n",
+    "closing": "Best Regards,\n",
+    "signature": "The Foundation\n\n"
+}
+
+
+def send_all():
+    """
+    Function to send a thank you letter to all donors.
+    This function sums total donations and writes letters to timestamped files.
+    """
+    for key, value in donors.items():
+       # I could not figure out how to get the timestamp to work with the filename for open(). I kept getting a file does not exist error.
+       # timestamp = datetime.now().strftime('%D')
+       # filename = "{}.txt.".format(key) + str(timestamp)
+        filename = "{}.txt".format(key)
+        file_content = '{greeting}' '{body}' '{closing}' '{signature}'.format(**email).format(key, sum(value))
+        files = open(filename, "w")
+        files.write(file_content)
+        print(str(files) + "  ---File created.\n")
+
+
+def create_report():
+    """
+    This function prints out a report with the following parameters:
+    Donor Name, Total Given, Number of Gifts, Average Gift Amount
+    """
+    header = ("Donor Name", "| Total Given", "| Num Gifts", "| Average Gift")
+    row = " ".join(["{:20s} {:>20s} {:>20s} {:>20s}"]).format(*header)
+    header_length = len(row)
+    print("\n" + row)
+    print("=" * header_length)
+    for key, value in sorted(donors.items()):
+        value_sum = str(sum(value))
+        value_len = str(len(value))
+        value_ave = str(sum(value)/(len(value)))
```

Comment:
you don't need to cast them as `str` before using in a string

Reply:
If I don't convert them to strings, I get exceptions like this.

Traceback (most recent call last):
  File "./mailroom.py", line 147, in <module>
    main()
  File "./mailroom.py", line 139, in main
    main_menu.get(user_input)()
  File "./mailroom.py", line 54, in create_report
    row_format = (key, "$" + value_sum, value_len, "$" + value_ave)
TypeError: can only concatenate str (not "float") to str